### ISSUE TITLE
Update to ndarray 0.14

### ISSUE THREAD

**Issue by relf** [rust-ml/linfa#109]
Just to mention I've started to look at using ndarray 0.14 (and hopefully I intend to make a PR in the end)

For now, I am blocked by an issue with `sprs` which is used by `linfa-kernel`: https://github.com/vbarrielle/sprs/issues/278

**Comment by bytesnake** [comment]
thanks for tackling this, last time I looked `ndarray-stats` still used the old `ndarray` version

**Comment by bytesnake** [comment]
but it isn't anymore, version 0.4.0 is already published on `crates.io`!

**Comment by relf** [comment]
ah yes! :facepalm: what did I check?

**Comment by quietlychris** [comment]
@relf As of two days ago (March 25) it actually looks like `ndarray` has bumped its version up from `0.14` to `0.15`. I'm not sure how many breaking chances are actually between the two, but I think if you're planning on updating the version up from `0.13` anyway, it might make sense to jump all the way up (assuming that there's no major issues in the `0.15.0`).

**Comment by relf** [comment]
Yep, I haven't dive in but it seems there is more changes from `0.14` to  `0.15` than from  `0.13` to `0.14`. 

That said, I think `linfa` should have at least a minor release for each minor release of `ndarray` (here we could have `linfa 0.4` with `ndarray 0.14`) because upgrading `ndarray` is not that easy due to the dependencies.

I do not know how cargo handles several versions of the same crate but I experienced compilation problems when there are various versions of `ndarray` drawn by various dependencies. And finally I waited that all dependencies upgraded to `ndarray 0.14` before making this PR.

Unfortunately, the `sprs` issue is blocking right now, and I do not know what is the best course of actions. I picked the revision vbarrielle/sprs@761d4f0f877de846b41b6e410822ce521ed008e8 where `sprs` upgraded to `ndarray 0.14` to make this PR compile. Should we wait for the issue resolution (which would be the same with `ndarray 0.15` anyway) or could we publish a `linfa 0.4` based on this?
